src/fs/filesystem.ts

```diff
@@ -165,13 +171,16 @@ export class FileSystem {
 
     const cid = await this.root.put()
 
+    // What are we using these hooks for?
     this.syncHooks.forEach(hook => {
```

Comment:
@dholms I don't actually see these used anywhere, aside from adding and removing them. Do we need this code?

Reply:
Yup for users to hook into. Steven uses sync hooks in drive

Reply:
Interesting. Any idea for what?

Reply:
Like, not sure if Drive should care if the files are coming off the network, right? Maybe something we can later clean up with a `Session` object?

Reply:
Here's the sync hook that Drive uses:
```js
async function syncHook(cid) {
  // stores the root cid locally, used as a cache, among other things.
  // a time difference currently determines when to use cache, or when to use sdk.dataRoot("username")
  // (see below)
  app.ports.ipfsReplaceResolvedAddress.send({ cid })

  // lastFsOperation determines when to use cache or when to use sdk.dataRoot("username")
  // → used on app start
  localStorage.setItem("fissionDrive.lastFsOperation", Date.now().toString())

  // update the user's data root
  await sdk.updateDataRoot(cid, {
    apiEndpoint: endpoints.api
  })
}
```

So yeah, used for "caching" local-first strategy and setting the user's data root. I also figured some users may wanted to hook into this, in case they wanted to do other stuff with the root CID of the file system.

> Like, not sure if Drive should care if the files are coming off the network, right?

It shouldn't care no. Not sure what it has to do with this though, can you clarify?

Reply:
> So yeah, used for "caching" local-first strategy and setting the user's data root. 

I'm very confused then. Doesn't the user need to call `Fission.sync()` (or something)? They'll get a promise, and know when it's resolved, along with a bunch of inspect-able data. What am I missing?

> It shouldn't care no. Not sure what it has to do with this though, can you clarify?

I'm clearly still confused about the use case. My understanding is that there are hooks for when the CID gets pushed or pulled from the remote. Not sure why the user would need to hook into that,  it feels very leaky, and my intuition is telling me that there's a cleaner design. But again, I clearly don't understand the use case fully.

Reply:
> Doesn't the user need to call Fission.sync() (or something)?

No, the sync hooks are called automatically when something changes in the file system.
For example, you do `fs.add()` and then after that it'll call the sync hook for you.

> My understanding is that there are hooks for when the CID gets pushed or pulled from the remote.

Like explained above, that's not what sync hooks are for. As shown in my Drive sync-hook example, the user is currently in total control over the data-root updating (that's what you meant by "push-pull from the remote", right?)

Reply:
> For example, you do fs.add() and then after that it'll call the sync hook for you.

This still feels very heavy and leaky. Lets do a call 😅

Reply:
From the call: 

* There probably aren't many other use cases for these hooks. Let's bake this in.
  * Debouncing and whatnot is likely going to get a lot easier when we have versioning
    * Because then we'll have Merkle proofs for which is the latest version
  * Default `add` probably should hit the server (so users don't forget)
    * Need to then either make a `addLocal` method, or add options `add(foo, {remote: false})`
      * My gut is telling me the config object
      * We're going to have ephemeral P2P updates and stuff, so _probably_ easier that way
      * Further thinking required, though
* Configuration (e.g. HTTP endpoint) is necessary
  * Either a Config object, or part of the `Fission.Session`